tailnet/conn_test.go

```diff
@@ -81,6 +81,83 @@ func TestTailnet(t *testing.T) {
 		_ = nc.Close()
 		<-conn
 
+		nodes := make(chan *tailnet.Node, 1)
+		w2.SetNodeCallback(func(node *tailnet.Node) {
+			select {
+			case nodes <- node:
+			default:
+			}
+		})
+		node := <-nodes
+		// Ensure this connected over DERP!
+		require.Len(t, node.DERPForcedWebsocket, 0)
+
+		w1.Close()
+		w2.Close()
+	})
+
+	t.Run("ForcesWebSockets", func(t *testing.T) {
+		t.Parallel()
+		w1IP := tailnet.IP()
+		derpMap := tailnettest.RunDERPOnlyWebSockets(t)
+		w1, err := tailnet.NewConn(&tailnet.Options{
+			Addresses:      []netip.Prefix{netip.PrefixFrom(w1IP, 128)},
+			Logger:         logger.Named("w1"),
+			DERPMap:        derpMap,
+			BlockEndpoints: true,
+		})
+		require.NoError(t, err)
+
+		w2, err := tailnet.NewConn(&tailnet.Options{
+			Addresses:      []netip.Prefix{netip.PrefixFrom(tailnet.IP(), 128)},
+			Logger:         logger.Named("w2"),
+			DERPMap:        derpMap,
+			BlockEndpoints: true,
+		})
+		require.NoError(t, err)
+		t.Cleanup(func() {
+			_ = w1.Close()
+			_ = w2.Close()
+		})
+		w1.SetNodeCallback(func(node *tailnet.Node) {
+			err := w2.UpdateNodes([]*tailnet.Node{node}, false)
+			assert.NoError(t, err)
+		})
+		w2.SetNodeCallback(func(node *tailnet.Node) {
+			err := w1.UpdateNodes([]*tailnet.Node{node}, false)
+			assert.NoError(t, err)
+		})
+		require.True(t, w2.AwaitReachable(context.Background(), w1IP))
+		conn := make(chan struct{})
+		go func() {
+			listener, err := w1.Listen("tcp", ":35565")
+			assert.NoError(t, err)
+			defer listener.Close()
+			nc, err := listener.Accept()
+			if !assert.NoError(t, err) {
+				return
+			}
+			_ = nc.Close()
+			conn <- struct{}{}
+		}()
+
+		nc, err := w2.DialContextTCP(context.Background(), netip.AddrPortFrom(w1IP, 35565))
+		require.NoError(t, err)
+		_ = nc.Close()
+		<-conn
+
+		nodes := make(chan *tailnet.Node, 1)
+		w2.SetNodeCallback(func(node *tailnet.Node) {
+			select {
+			case nodes <- node:
+			default:
+			}
+		})
+		node := <-nodes
+		require.Len(t, node.DERPForcedWebsocket, 1)
+		// Ensure the reason is valid!
+		require.Equal(t, "GET failed with status code 400: Invalid \"Upgrade\" header: DERP", node.DERPForcedWebsocket[derpMap.RegionIDs()[0]])
```

Comment:
You should also test that websocket derps work here too, by updating the handler to only accept derp if Upgrade: websocket.